File: LC/[0133] Clone Graphy.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    unordered_map<Node *, Node *> old2new;
    Node* cloneGraph(Node* node) {
        if(node == nullptr) return nullptr;

        if(old2new.find(node) != old2new.end()){
            return old2new[node];//already cloned (visited)
        }

        old2new[node]=new Node(node->val); //make a copy
        for(auto & c: node->neighbors){ //DFS for all neighbors
            old2new[node]->neighbors.push_back(cloneGraph(c)); //adjacency list
        }

        return old2new[node];
    }
};

File: LC/[0133] Clone Graphy.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None: return None

        old2new={}
        def dfs(node):
            if node is None: return None
            if node in old2new: return old2new[node]

            old2new[node]=Node(node.val)

            for child in node.neighbors:
                old2new[node].neighbors.append(dfs(child))

            return old2new[node]


        dfs(node)

        return old2new[node]


####################################
class Solution:    
    def __init__(self):
        self.orig2new={}
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None: return None

        if node in self.orig2new: return self.orig2new[node]
        self.orig2new[node]=Node(node.val)

        for child in node.neighbors:
            self.orig2new[node].neighbors.append(self.cloneGraph(child))

        return self.orig2new[node]
